### PR TITLE
Adding Particle Extrema Reduced Diagnostics

### PR DESCRIPTION
**Pull request by Yin-YinjianZhao**
This PR adds a reduced diagnostic that outputs the max/min values of some particle information.
This PR is motivated by issue #1311 .

The outputs are:
```
xmin, xmax, ymin, ymax, zmin, zmax, pxmin, pxmax, pymin, pymax, pzmin, pzmax, gmin, gmax, wmin, wmax.
```
for regular species.

When `XZ` dimension is used, `ymin, ymax` are zeros.
When `RZ` dimension is used, `x = r*cos(theta), y = r*sin(theta)` are used to still obtain `xmin, xmax, ymin, ymax`.

### CODE REVIEW

**Comment by Yin-YinjianZhao** [comment]
Testing input scripts are:
[inputrz.txt](https://github.com/ECP-WarpX/WarpX/files/5211526/inputrz.txt)
[input3d.txt](https://github.com/ECP-WarpX/WarpX/files/5211527/input3d.txt)
[input2d.txt](https://github.com/ECP-WarpX/WarpX/files/5211528/input2d.txt)
The outputed `E.dat` is compared with the `diags` using the scripts:
[plotrz.txt](https://github.com/ECP-WarpX/WarpX/files/5211534/plotrz.txt)
[plot3d.txt](https://github.com/ECP-WarpX/WarpX/files/5211535/plot3d.txt)
[plot2d.txt](https://github.com/ECP-WarpX/WarpX/files/5211536/plot2d.txt)
They match each other.

**Comment by Yin-YinjianZhao** [comment]
@lucafedeli88 Hi Luca, I may ready to add special species. One question is how to judge if a species is a special species. I saw you did `if (myspc.AmIA<PhysicalSpecies::photon)` in `ParticleEnergy.cpp`. Maybe we can do similar things. Another question, for the QED species, what's the formula to compute x? I'm not familiar with QED stuff.

**Comment by lucafedeli88** [comment]
> @lucafedeli88 Hi Luca, I may ready to add special species. One question is how to judge if a species is a special species. I saw you did `if (myspc.AmIA<PhysicalSpecies::photon)` in `ParticleEnergy.cpp`. Maybe we can do similar things. Another question, for the QED species, what's the formula to compute x? I'm not familiar with QED stuff.

Hi @Yin-YinjianZhao ! First of all, thanks a lot for this PR !

As far as your question is of concern, a possible option is to calculate the χ parameter only for species with QED processes enabled. You could test if a species has the Breit-Wheeler (BW) process enabled (`pc->has_breit_wheeler()`) or the Quantum Synchrotron (QS) process enabled  (`pc->has_quantum_sync()`). Only photons can have BW enabled and only electrons and positrons can have QS enabled, however `has_quantum_sync` and `has_breit_wheeler` are member functions of the parent class `WarpXParticleContainer`. So you could do something like:
```
    for (const auto& pc : allcontainers){
        if(!pc->has_breit_wheeler() && !pc->has_quantum_sync())
            continue;
        /* CALCULATE MIN AND MAX CHI PARAMETER */
     }
``` 

In order to calculate the χ parameter you can use the functions `chi_photon` and `chi_lepton` available in `WarpX/Source/Particles/ElementaryProcess/QEDInternals/QedChiFunctions.H` . These functions are thin wrappers around the corresponding functions provided by PICSAR. Unfortunately, they require both particle momenta and electromagnetic fields.

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+                if ( myspc.AmIA<PhysicalSpecies::photon>() )
+                { Real chi = chi_photon(px, py, pz, ex, ey, ez, bx, by, bz); }
+                else
+                { Real chi = chi_lepton(px, py, pz, ex, ey, ez, bx, by, bz); }
+
+                return 0.0_rt;
+            });
+            ParallelDescriptor::ReduceRealMin(chimin);
+        }
+#endif
```
Comment (Yin-YinjianZhao) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: @atmyers @lucafedeli88 Hi Andrew and Luca, does this part of the code look correct?

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+            {
+#if (defined WARPX_DIM_RZ)
+                ParticleReal xp = p.pos(0)*std::cos(p.rdata(PIdx::theta));
+                ParticleReal yp = p.pos(0)*std::sin(p.rdata(PIdx::theta));
+#elif (defined WARPX_DIM_XZ)
+                ParticleReal xp = p.pos(0);
+                ParticleReal yp = 0.0_rt;
+#else
+                ParticleReal xp = p.pos(0);
+                ParticleReal yp = p.pos(1);
```
Comment (atmyers) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: Could you use the GetPosition functor to do this?

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+#endif
+                ParticleReal zp = p.pos(index_z);
+                ParticleReal ex = 0._rt, ey = 0._rt, ez = 0._rt;
+                ParticleReal bx = 0._rt, by = 0._rt, bz = 0._rt;
```
Comment (atmyers) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: It looks like we aren't taking external fields into account here.

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp (hunk 1 above)
Comment (lucafedeli88) [reply Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: Great! 
I just have a couple of comments:
- I agree with @atmyers that the external fields should be taken into account. To this end I guess that you should generate these two functors `auto get_externalE = GetExternalEField  (pti, offset);` and `auto get_externalB = GetExternalBField  (pti, offset);` just before the for loop. Then, before calling `doGatherShapeN`, you can do:
```
                    ParticleReal ex = 0._rt, ey = 0._rt, ez = 0._rt;
                    get_externalE(i, ex, ey, ez);
                    ParticleReal bx = 0._rt, by = 0._rt, bz = 0._rt;
                    get_externalB(i, bx, by, bz);
```  
- Shouldn't the functor inside `ReduceMin` return `chi` ?

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+                IndexType bx_type, by_type, bz_type;
+                GpuArray<amrex::Real, 3> dx_arr, xyzmin_arr;
+                Dim3 lo;
+
+                int n_rz_azimuthal_modes;
+                int nox;
+                bool galerkin_interpolation;
+
+                get_externalE(i, ex, ey, ez);
+                get_externalB(i, bx, by, bz);
```
Comment (Yin-YinjianZhao) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: @atmyers @lucafedeli88 
How can I get `i` in this `ReduceMax/Min` loop for `get_externalE(i, ex, ey, ez)` etc.
Something like?
```
chimin = ReduceMin( myspc,
            [=] AMREX_GPU_HOST_DEVICE (const PType& p, int i) -> Real
            { ... }
```

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+    }
+}
+// end constructor
+
+// function that computes extrema
+void ParticleExtrema::ComputeDiags (int step)
+{
+
+    // Judge if the diags should be done
+    if ( (step+1) % m_freq != 0 ) { return; }
```
Comment (NeilZaim) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]:
```suggestion
    if (!m_intervals.contains(step+1)) { return; }
```

Hello! Just a minor modification that you'll have to make since we've implemented the intervals parser for reduced diagnostics.

Comment (Yin-YinjianZhao) [reply Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: @NeilZaim Thanks! Did that.

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+                    chimin[lev] = get<0>(reduce_data.value());
+                    chimax[lev] = get<1>(reduce_data.value());
+                }
+                chimin_f = std::min_element(chimin.begin(), chimin.end());
+                chimax_f = std::max_element(chimax.begin(), chimax.end());
+            }
+            ParallelDescriptor::ReduceRealMin(chimin_f);
+            ParallelDescriptor::ReduceRealMax(chimax_f);
+        }
+#endif
```
Comment (Yin-YinjianZhao) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: @atmyers Hi Andrew, does this part of the code look correct?

Comment (Yin-YinjianZhao) [reply Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: @lucafedeli88 Hi Luca, do you think this part of the code is correct?

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+{
+
+    // Judge if the diags should be done
+    if (!m_intervals.contains(step+1)) { return; }
+
+    // get MultiParticleContainer class object
+    auto & mypc = WarpX::GetInstance().GetPartContainer();
+
+    // get number of level (int)
+    auto level_number = WarpX::GetInstance().finestLevel();
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]:
```suggestion
    const auto level_number = WarpX::GetInstance().finestLevel();
```
I think that this could be `const`

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+
+    // get number of species (int)
+    auto nSpecies = mypc.nSpecies();
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]:
```suggestion
    const auto nSpecies = mypc.nSpecies();
```
I think that also this could be `const`

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+    pp.get("species",m_species_name);
+
+    // get WarpX class object
+    auto & warpx = WarpX::GetInstance();
+
+    // get MultiParticleContainer class object
+    auto & mypc = warpx.GetPartContainer();
+
+    // get number of species (int)
+    auto nSpecies = mypc.nSpecies();
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]:
```suggestion
    const auto nSpecies = mypc.nSpecies();
```
I think that this could be `const`

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+
+    // get species names (std::vector<std::string>)
+    auto species_names = mypc.GetSpeciesNames();
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]:
```suggestion
    const auto species_names = mypc.GetSpeciesNames();
```
I think that also this could be `const`

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+        // only chosen species does
+        if (species_names[i_s] != m_species_name) { continue; }
+
+        // get WarpXParticleContainer class object
+        auto & myspc = mypc.GetParticleContainer(i_s);
+
+        if (myspc.has_breit_wheeler() || myspc.has_quantum_sync())
+        {
+            // resize data array for QED species
+            m_data.resize(18,0.0);
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]:
```suggestion
            const int num_quantities = 18;
            m_data.resize(num_quantities,0.0);
```
I would suggest to write something like this to avoid unnamed "magic" numbers!

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+        } else
+        {
+            // resize data array for regular species
+            m_data.resize(16,0.0);
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]:
```suggestion
            const int num_quantities = 16;
            m_data.resize(num_quantities,0.0);
```
same as above

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+            {
+                // open file
+                std::ofstream ofs;
+                ofs.open(m_path + m_rd_name + "." + m_extension,
+                    std::ofstream::out | std::ofstream::app);
+                // write header row
+                ofs << "#";
+                ofs << "[1]step()";
+                ofs << "[2]time(s)";
+                ofs << "[3]xmin()";
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: Why there is no `ofs << m_sep` between these?

Just if you want, you can make this section more compact by doing something like:
```
auto header = std::vector<string>{"[1]step()", "[2]time(s)", ...};
 if (myspc.has_breit_wheeler() || myspc.has_quantum_sync()){
    header.push_back("[19]chimin()");
    header.push_back("[20]chimax()");
}
for (const auto& entry : header){
    ofs << entry << m_sep;
}
```

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+    auto & mypc = WarpX::GetInstance().GetPartContainer();
+
+    // get number of level (int)
+    auto level_number = WarpX::GetInstance().finestLevel();
+
+    // get number of species (int)
+    auto nSpecies = mypc.nSpecies();
+
+    // get species names (std::vector<std::string>)
+    auto species_names = mypc.GetSpeciesNames();
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]:
```suggestion
    const auto species_names = mypc.GetSpeciesNames();
```
also this could be `const`

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 15]
```diff
@@ -0,0 +1,419 @@
...
+    {
+
+        // only chosen species does
+        if (species_names[i_s] != m_species_name) { continue; }
+
+        // get WarpXParticleContainer class object
+        auto & myspc = mypc.GetParticleContainer(i_s);
+
+        // get mass (Real)
+        auto m = myspc.getMass();
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp:173]:
```suggestion
        const auto m = myspc.getMass();
```
I think that this could be `const`

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 16]
```diff
@@ -0,0 +1,419 @@
...
+
+        using PType = typename WarpXParticleContainer::SuperParticleType;
+
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp:180]: If you want to make this part slightly more compact you could consider something like
(I think that you can also omit `->Real`):
```
auto get_x =  
#if (defined WARPX_DIM_RZ)
       [=] AMREX_GPU_HOST_DEVICE (const PType& p) {
            return p.pos(0)*std::cos(p.rdata(PIdx::theta));};
#else
       [=] AMREX_GPU_HOST_DEVICE (const PType& p) {return p.pos(0);};
#endif
```
If you do this, you can just write: 
```
Real xmin = ReduceMin( myspc, get_x);
ParallelDescriptor::ReduceRealMin(xmin);
```

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 17]
```diff
@@ -0,0 +1,419 @@
...
+                    // declare position arrays
+                    Gpu::ManagedDeviceVector<ParticleReal> x, y, z;
+                    // get particle arrays
+                    pti.GetPosition(x, y, z);
+                    auto & w  = pti.GetAttribs(PIdx::w);
+                    auto & ux = pti.GetAttribs(PIdx::ux);
+                    auto & uy = pti.GetAttribs(PIdx::uy);
+                    auto & uz = pti.GetAttribs(PIdx::uz);
+                    // declare external fields
+                    int offset = 0;
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]:
```suggestion
                    const int offset = 0;
```
Could this be `const` ?

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 18]
```diff
@@ -0,0 +1,419 @@
...
+                    const auto getExternalE = GetExternalEField(pti, offset);
+                    const auto getExternalB = GetExternalBField(pti, offset);
+                    // declare reduce_op
+                    ReduceOps<ReduceOpMin, ReduceOpMax> reduce_op;
+                    ReduceData<Real, Real> reduce_data(reduce_op);
+                    using ReduceTuple = typename decltype(reduce_data)::Type;
+                    reduce_op.eval(ux.size(), reduce_data,
+                    [=] AMREX_GPU_DEVICE (int i) -> ReduceTuple
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp:382]: I think that `->ReduceTuple`  could be omitted here

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 19]
```diff
@@ -0,0 +1,419 @@
...
+                            ex, ey, ez, bx, by, bz,
+                            ex_arr, ey_arr, ez_arr, bx_arr, by_arr, bz_arr,
+                            ex_type, ey_type, ez_type,
+                            bx_type, by_type, bz_type,
+                            dx_arr, xyzmin_arr, lo,
+                            n_rz_azimuthal_modes, nox, galerkin_interpolation);
+                        // compute chi
+                        Real chi = 0.0_rt;
+                        if ( myspc.AmIA<PhysicalSpecies::photon>() ) {
+                            chi = chi_photon(ux[i]*m, uy[i]*m, uz[i]*m,
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: There is actually an issue with this line: `m` for photons is 0! `ux`, `uy` and `uz` for photons are calculated as if they had the same mass of an electron. This should probably be better emphasized in the developers documentation.
Since only photons can have this QED process, you can simply replace m with the electron mass (and maybe add a one line comment)

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 20]
```diff
@@ -0,0 +1,419 @@
...
+
+        // gmin
+        Real gmin = ReduceMin( myspc,
+        [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+        {
+            Real ux = p.rdata(PIdx::ux);
+            Real uy = p.rdata(PIdx::uy);
+            Real uz = p.rdata(PIdx::uz);
+            Real us = ux*ux + uy*uy + uz*uz;
+            return std::sqrt(1.0 + us*inv_c2);
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: This should be different for photons, since their mass is zero. Quite often people define a sort of `gamma` factor for photons, which is simply their momentum normalized with respect to m_e c.
You could simply replace `std::sqrt(1.0 + us*inv_c2);` with `std::sqrt(us*inv_c2);` for photons

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 21]
```diff
@@ -0,0 +1,419 @@
...
+
+        // gmax
+        Real gmax = ReduceMax( myspc,
+        [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+        {
+            Real ux = p.rdata(PIdx::ux);
+            Real uy = p.rdata(PIdx::uy);
+            Real uz = p.rdata(PIdx::uz);
+            Real us = ux*ux + uy*uy + uz*uz;
+            return std::sqrt(1.0 + us*inv_c2);
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp]: same as above

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 22]
```diff
@@ -0,0 +1,419 @@
...
+        m_data[2]  = ymin;
+        m_data[3]  = ymax;
+        m_data[4]  = zmin;
+        m_data[5]  = zmax;
+        m_data[6]  = uxmin*m;
+        m_data[7]  = uxmax*m;
+        m_data[8]  = uymin*m;
+        m_data[9]  = uymax*m;
+        m_data[10] = uzmin*m;
+        m_data[11] = uzmax*m;
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp:441]: Here `m` should be replaced with the electron mass for photons

**Review by lucafedeli88 — CHANGES_REQUESTED** [review]
Thanks a lot for this PR!  I think that there is only one issue: photons have zero mass and ux, uy, uz are calculated as if they had a mass equal to the electron mass. Therefore, some small changes are required to take this into account(I left some comments). I also left few other minor comments (they can be safely ignored).

**Comment by lucafedeli88** [comment]
The test `particles_in_pml` has failed:
```
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Ex]
Benchmark: [lev=0,Ex] 63840.5064701782321208156645298004150390625000
Plotfile : [lev=0,Ex] 63840.5063970984192565083503723144531250000000

   execution time: 52.5s
   creating problem test report ...
   particles_in_pml FAILED
```
I don't think that this is due to this PR. So I would suggest to simply to merge with `upstream\development` (or to rebase on `upstream\development`) and see if the tests pass this time.

**Comment by Yin-YinjianZhao** [comment]
@lucafedeli88 Hi Luca, do we need an automated test? I'm not very familiar with QED simulations, maybe you can tested it using the simulation that you needed to look at the extrema values previously, or provide me an input script, or maybe we can use an example QED run?

**Comment by lucafedeli88** [comment]
> @lucafedeli88 Hi Luca, do we need an automated test? I'm not very familiar with QED simulations, maybe you can tested it using the simulation that you needed to look at the extrema values previously, or provide me an input script, or maybe we can use an example QED run?

Hi @Yin-YinjianZhao ! Sorry for my late answer... I've tested your PR and it seems to me that it is working nicely! For the automated tests, there are two options: either we add a few lines to the QED tests (that's how I tested your PR on my machine) or we wait for https://github.com/ECP-WarpX/WarpX/pull/1486 and we add a few lines to the tests that you have prepared for this PR. If you want, I can take care of that in a separate PR...

**Comment by Yin-YinjianZhao** [comment]
> > @lucafedeli88 Hi Luca, do we need an automated test? I'm not very familiar with QED simulations, maybe you can tested it using the simulation that you needed to look at the extrema values previously, or provide me an input script, or maybe we can use an example QED run?
> 
> Hi @Yin-YinjianZhao ! Sorry for my late answer... I've tested your PR and it seems to me that it is working nicely! For the automated tests, there are two options: either we add a few lines to the QED tests (that's how I tested your PR on my machine) or we wait for #1486 and we add a few lines to the tests that you have prepared for this PR. If you want, I can take care of that in a separate PR...

Hi @lucafedeli88 , thanks for your reply! I'm fine with either way. Since I haven't done any tests of QED, it would be much helpful if you can take care of adding the automated test, maybe just using the one that you tested. It's okay either in this PR, or in a separate PR. Thank you so much!

**Comment by lucafedeli88** [comment]
> > > @lucafedeli88 Hi Luca, do we need an automated test? I'm not very familiar with QED simulations, maybe you can tested it using the simulation that you needed to look at the extrema values previously, or provide me an input script, or maybe we can use an example QED run?
> > 
> > 
> > Hi @Yin-YinjianZhao ! Sorry for my late answer... I've tested your PR and it seems to me that it is working nicely! For the automated tests, there are two options: either we add a few lines to the QED tests (that's how I tested your PR on my machine) or we wait for #1486 and we add a few lines to the tests that you have prepared for this PR. If you want, I can take care of that in a separate PR...
> 
> Hi @lucafedeli88 , thanks for your reply! I'm fine with either way. Since I haven't done any tests of QED, it would be much helpful if you can take care of adding the automated test, maybe just using the one that you tested. It's okay either in this PR, or in a separate PR. Thank you so much!

If it's okay for you then, I will open a separate PR as soon as this one is merged...

File: Source/Diagnostics/ReducedDiags/MultiReducedDiags.cpp [hunk 23]
```diff
@@ -81,6 +82,11 @@ MultiReducedDiags::MultiReducedDiags ()
             m_multi_rd[i_rd]=
                 std::make_unique<ParticleNumber>(m_rd_names[i_rd]);
         }
+        else if (rd_type.compare("ParticleExtrema") == 0)
+        {
+            m_multi_rd[i_rd].reset
+                ( new ParticleExtrema(m_rd_names[i_rd]));
```
Comment (ax3l) [inline Source/Diagnostics/ReducedDiags/MultiReducedDiags.cpp]: Just for consistency with the ones around it and modernization:
```suggestion
            m_multi_rd[i_rd] = 
                std::make_unique< ParticleExtrema >(m_rd_names[i_rd]);
```

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.H [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+    /** constructor
+     *  @param[in] rd_name reduced diags names */
+    ParticleExtrema(std::string rd_name);
+
+    /// name of species
+    std::string m_species_name;
+
+    /** This funciton computes the particle extrema
+     */
+    virtual void ComputeDiags(int step) override final;
```
Comment (ax3l) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.H]: Only `override` in that case:
```suggestion
    void ComputeDiags(int step) override final;
```
Ref.: https://stackoverflow.com/questions/39932391/virtual-override-or-both-c

**Review by ax3l — COMMENTED** [review]
Looks great, I added a few minor comments and if @NeilZaim and @lucafedeli88 are also :+1: will merge it then.

**Comment by NeilZaim** [comment]
Hello! I didn't read the PR in detail but I have a couple of general questions/comments:

- Do you want to update the parameters documentation here or in a separate PR (e.g. once we have automated tests)?

- If I understand correctly, if the species has QED effects enabled, we will by default perform another field gathering to compute the extrema of chi. Is this a bit dangerous? E.g. if a user is unaware of that and uses the default reduced diagnostic period of 1, is there a risk of starting to have significant performance losses? I guess we can add a warning in the parameters documentation. We could also consider adding another optional input parameters `list_of_quantities` where the user can choose which quantities he wants to get the extrema of, with Chi turned off by default. (if we could actually do the reduction at the same time as the field gathering it would be neater but I guess that would make things more complex) What do you think about this?

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 25]
```diff
@@ -0,0 +1,452 @@
...
+                ofs << m_sep;
+                ofs << "[14]pzmax(kg*m/s)";
+                ofs << m_sep;
+                ofs << "[15]gmin()";
+                ofs << m_sep;
+                ofs << "[16]gmax()";
+                ofs << m_sep;
+                ofs << "[17]wmin()";
+                ofs << m_sep;
+                ofs << "[18]wmax()";
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp:111]: if DIM=2 these quantities have actually dimension `1/m`

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp [hunk 26]
```diff
@@ -0,0 +1,452 @@
...
+        { return p.rdata(PIdx::ux); });
+        ParallelDescriptor::ReduceRealMin(uxmin);
+
+        // uxmax
+        Real uxmax = ReduceMax( myspc,
+        [=] AMREX_GPU_HOST_DEVICE (const PType& p)
+        { return p.rdata(PIdx::ux); });
+        ParallelDescriptor::ReduceRealMax(uxmax);
+
+        // uymin
```
Comment (lucafedeli88) [inline Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp:261]: I was wondering: are all those calls to `ReduceMin` and `ReduceMax` an issue? Would it be better to group all them in a `reduce_op.eval` loop ?
(if this is the case we might do that in another PR)

**Review by lucafedeli88 — COMMENTED** [review]
It looks generally good! I've just left a couple of small comments in addition to Neil's ones...

**Comment by Yin-YinjianZhao** [comment]
> Hello! I didn't read the PR in detail but I have a couple of general questions/comments:
> 
> * Do you want to update the parameters documentation here or in a separate PR (e.g. once we have automated tests)?
> * If I understand correctly, if the species has QED effects enabled, we will by default perform another field gathering to compute the extrema of chi. Is this a bit dangerous? E.g. if a user is unaware of that and uses the default reduced diagnostic period of 1, is there a risk of starting to have significant performance losses? I guess we can add a warning in the parameters documentation. We could also consider adding another optional input parameters `list_of_quantities` where the user can choose which quantities he wants to get the extrema of, with Chi turned off by default. (if we could actually do the reduction at the same time as the field gathering it would be neater but I guess that would make things more complex) What do you think about this?

Thanks for the comments! I added the doc, and mentioned the performance may be bad in the doc for now.

**Review by NeilZaim — APPROVED** [review]
Thanks adding the docs and thanks again for the new diag! On my side I think this PR can be merged.

**Comment by ax3l** [comment]
@Yin-YinjianZhao  can you please rebase against the latest `development` or merge that branch in?

@lucafedeli88 is the PR :+1: from your side, too? :)

File: Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp (hunk 26 above)
Comment (ax3l) [reply Source/Diagnostics/ReducedDiags/ParticleExtrema.cpp:261]: Great idea! They are probably not a severe issue per se, but grouping them would definitely be a nice follow-up PR to speed up the diagnostics (latency and performance at scale).

**Review by lucafedeli88 — APPROVED** [review]
Yes, I approve this PR!

**Review by ax3l — APPROVED** [review]
Thank you @Yin-YinjianZhao and thank you @NeilZaim and @lucafedeli88 for the great review work! :sparkles:
